src/core/or/connection_edge.c

```diff
@@ -863,18 +863,35 @@ export_hs_client_circuit_id(edge_connection_t *edge_conn,
   if (protocol != HS_CIRCUIT_ID_PROTOCOL_HAPROXY)
     return;
 
+  origin_circuit_t *circ;
   char *buf = NULL;
   const char dst_ipv6[] = "::1";
   /* See RFC4193 regarding fc00::/7 */
-  const char src_ipv6_prefix[] = "fc00:dead:beef:4dad:";
+  const char src_ipv6_prefix[] = "fc00:";
   uint16_t dst_port = 0;
   uint16_t src_port = 1; /* default value */
   uint32_t gid = 0; /* default value */
+  uint32_t rp_addr = 0;
+  uint16_t rp_port = 0;
 
   /* Generate a GID and source port for this client */
   if (edge_conn->on_circuit != NULL) {
-    gid = TO_ORIGIN_CIRCUIT(edge_conn->on_circuit)->global_identifier;
+    circ = TO_ORIGIN_CIRCUIT(edge_conn->on_circuit);
+    gid = circ->global_identifier;
     src_port = gid & 0x0000ffff;
+    if (get_options()->HiddenServiceExportRendPoint) {
+      crypt_path_t *cpath = circ->cpath;
+      crypt_path_t *cpath_last = NULL;
+      if (cpath) {
+        for (; cpath_last != cpath; cpath = cpath_last) {
+          cpath_last = cpath->next;
+        }
```

Comment:
So I believe, on an `origin_circuit_t`, you can avoid going through the cpath and just use the build state:

`circ->build_state->chosen_exit`

Which is conveniently the `extend_info_t` you need below. Just be careful, the `build_state` could be NULL but unlikely so you could use your macro `BUG()` for that and bail if so.

Reply:
Fixed in `0599c72`.

Reply:
Oh I think you can simply remove the fallback through the `cpath` and just bail out after the BUG() with a simple `return;`.

The point is that it is really not suppose to happen but if so, we'll notice and we'll be able to fix it. Chances are that the `cpath` is broken if the `chosen_exit` is not present. Not only that but you might end up using a `cpath` that is not at all the rendezvous node for instance.

Reply:
Fixed in `a932db5`.